[ENH] `evaluate` for TSC and TSR

jgyasu opened the pull request:
Implements `evaluate` utility for benchmarking time series classifiers and regressors.

This PR branches off #6888 and builds on it. 

Fixes #6852, fixes #6708

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,439 @@
+    return {}
+
+
+def _evaluate_window(x, meta):
```
maybe we should avoid calling this "window". This does a single run of cross-validation.

jgyasu replied:
How about calling it `_evaluate_fold`? Since every single iteration of CV is called  fold?

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,439 @@
+
+    Returns
+    -------
+     results : pd.DataFrame or dask.dataframe.DataFrame
```
looks like an extra space, accident

jgyasu replied:
I did not understand, extra space where exactly? On which line?

fkiraly replied:
there is a *superfluous* space before `results` which should be removed

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,439 @@
+
+         - y_test: (pd.Series) present if see ``return_data=True``
+         testing fold of the i-th split in ``cv``, used to compute the metric.
+    """
```
I would suggest adding an example, that you can manually test

jgyasu replied:
Sounds good, will do!

jgyasu commented:
The docstring example is failing, debugging is a WIP.

jgyasu commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,450 @@
+        """
+        splitter = InstanceSplitter(cv)
+
+        geny = splitter.split(y)
```
@fkiraly this wasn't caught in the reviews earlier but I believe the failing tests urrently are due to trying to split `y` into instances while it is just the label. I tried testing `InstanceSplitter` seperately and observed the same error,

```python
from sktime.split import InstanceSplitter
from sktime.datasets import load_unit_test
from sklearn.model_selection import KFold
X, y = load_unit_test()
splitter = InstanceSplitter(KFold(n_splits=3))
X_splits = splitter.split(X)
first_split = next(X_splits)
```
This works perfectly alright, but gives the same error if I use `split` on `y`

jgyasu commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,448 @@
+
+        genx = splitter.split_series(X)
+
+        yield from genx
```
from my understanding, `InstanceSplitter(cv).split_series(X)` first groups the data by instances and then split them into test and train

I wonder how we can get the same split for the labels `y`, I was looking at `apply_split` but it wasn't very clear to me

@fkiraly

jgyasu replied:
Will it nor make more sense to use `splitter.split(X)` instead? since it will return `iloc` and based on it we can also split `y`?

fkiraly replied:
yes, exactly. And that you use in `apply_split` twice.

I do not think the `InstanceSplitter` is helpful here.

fkiraly replied:
I do not think you need `InstanceSplitter` - simply use `apply_split` for both `X` and `y`

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,455 @@
+        X_train : i-th train split of y as per cv.
+        X_test : i-th test split of y as per cv.
+        """
+        splitter = InstanceSplitter(cv)
```
this does not look right - why do you not remove `InstanceSplitter` entirely?

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,455 @@
+
+
+def _get_column_order_and_datatype(
+    metric_types: dict, return_data: bool = True, cutoff_dtype=None, old_naming=True
```
this is copy-paste from forecasting - I think the default should be `old_naming=False`, and all logic related to `old_naming=True` should be removed.

fkiraly reviewed (changes requested):
Nice, it looks like the problem with `apply_map` is solved!

The current problem looks different - looking at the error logs, it seems some logic from forecasting metrics (wrapping in dynamic metrics etc) has been copied over, which is not necessary. Removing all that should solve the problem!

fkiraly commented:
> the tests will currently fail, because `_check_scores` returns a list and `_evaluate_fold` expects a dict, since earlier it was just a copy paste of TSF `evaluate` and used sktime's metrics with extractable scitypes 

Can you kindly be more specific which parameters must be a list/dict, etc, and how the passing works or worked?

An issue that we will still have is to ensure that probabilistic classification metrics get the output of `predict_proba`, while non-probabilistic ones get the output of `predict`.

jgyasu commented:
I have updated the `_check_scores` to be compatible for time series classification with sklearn metrics. Following @fkiraly's suggestions, the function now segragates metrics suited for probabilistic and deterministic function with the help of the `_is_proba_classification_score` helper function.

Still looking into the failing tests.

jgyasu commented:
Fixed some more copy-paste, now the result DataFrame contains `X_train` and `X_test` too. But I have removed the parallelize logic for now, I am trying to find why it outputs an empty structure.
Here's the current result from the docstring example run:
![Screenshot from 2025-05-29 20-11-04](https://github.com/user-attachments/assets/888c4d76-2ec0-4328-b2a7-56f21c7d9519)

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,490 @@
+    results = []
+    for x in enumerate(yx_splits):
+        result, classifier = _evaluate_fold(x, _evaluate_fold_kwargs)
+        _evaluate_fold_kwargs["classifier"] = classifier
```
why is this line necessary?

fkiraly replied:
but was it not already populated above?

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,490 @@
+            train_instances = instance_idx[train_instance_idx]
+            test_instances = instance_idx[test_instance_idx]
+
+            X_train = X.loc[X.index.get_level_values(0).isin(train_instances)]
```
why are we not using `apply_split`?

fkiraly replied:
I see - maybe then the function needs to be adapted so it is easier to use

fkiraly reviewed (changes requested):
Looks good - left some small comments above.

We should add `parallelize` in a PR stacking on top of this.

jgyasu commented on sktime/utils/multiindex.py:
```diff
@@ -1,7 +1,7 @@
 """Multiindex formatting related utilities."""
 
-__author__ = ["fkiraly"]
+__author__ = ["fkiraly", "ksharma6"]
```
Seems like `apply_split` function was written by them:

<img width="814" alt="Screenshot 2025-06-01 at 8 06 07 PM" src="https://github.com/user-attachments/assets/77a41702-ffeb-40c0-9a56-7702f52db527" />

fkiraly replied:
well, I wrote it in their PR. But let´s leave it in, they started it.

jgyasu replied:
I think it should not be there then - if it was your work. But sure!

fkiraly replied:
I think when in doubt, better be generous. Let's leave it there.

jgyasu commented:
Oops, I didn't realise that the result was not displaying `y_pred` and `pred_test`, made some changes to enable that but now it leads to another concern - estimator is not getting fitted. I am looking into the bug causing it.

jgyasu commented:
I have currently hardcoded positive label to be 1 for metrics which expects a `pos_label`. Please let me know if there is any better design.

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,481 @@
+    if metric_name in PROBA_METRICS:
+        return True
+
+    try:
```
why is a try/except needed here?

Can we try to replace it by if/else? It is best practice to avoid try/except by if/else using the exact failure condition if what we want to do is not actually error handling but type condition handling.

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,481 @@
+    except (ValueError, TypeError):
+        pass
+
+    try:
```
to contrast: this try/except is ok because it is a form of "detection by execution"

fkiraly commented on sktime/classification/model_evaluation/_functions.py:
```diff
@@ -0,0 +1,481 @@
+        Provides train/test indices to split data in train/test sets.
+       Splits the dataset into k consecutive folds (without shuffling by default).
+    X : sktime compatible time series panel data container, Panel scitype, e.g.,
+             pd-multiindex: pd.DataFrame with columns = variables,
```
this formatting will not properly render - can you look for a docstring where this is a bullet point list and renders properly?
